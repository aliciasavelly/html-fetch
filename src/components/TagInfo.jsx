import React, { Component } from 'react';

class TagInfo extends Component {
  constructor(props) {
    super(props);

    this.state = {
      tags: {},
    }

    this.renderTags = this.renderTags.bind(this);
  }

  componentWillReceiveProps(newProps) {
    if (newProps.newTags) {
      this.setState({ tags: newProps.tags });
    }
  }

  renderTags() {
    const { tags } = this.state;
    // debugger;
    if (Object.keys(tags).length == 0) {
      return(
        <div className="instructions">Enter the URL of a web page in the search bar above! The page's HTML will be loaded, and you can toggle which tags you'd like to be highlighted in the source code view.</div>
      )
    } else {
      return(
        <div className="tags-index">
          {this.props.sortedTags.map( (tag, idx) => (
            <p key={`tag-${idx}`} className="tag" onClick={this.props.handleMark}>{tag.toLowerCase()}: {tags[tag]}</p>
          ))}
        </div>
      )
    }
  }

  render() {
    return(
      <div>
        {this.renderTags()}
      </div>
    )
  }
}

export default TagInfo;
