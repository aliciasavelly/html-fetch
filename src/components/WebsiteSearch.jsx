import React, { Component } from 'react';
import axios from 'axios';
import TagInfo from './TagInfo';
import _ from 'lodash';
import { Mark } from 'mark.js';

class WebsiteSearch extends Component {
  constructor(props) {
    super(props);

    this.state = {
      website: '',
      tags: {},
      error: '',
      newTags: false,
      sortedTags: [],
      doc: {},
      regex: []
    }

    this.handleSubmit = this.handleSubmit.bind(this);
    this.updateWebsite = this.updateWebsite.bind(this);
    this.addTags = this.addTags.bind(this);
    this.handleMark = this.handleMark.bind(this);
  }

  handleSubmit(e) {
    e.preventDefault();

    let website = this.state.website;
    const self = this;

    if (website.slice(0, 3) === "www") {
      website = "https://" + website;
    }

    axios.get(`http://cors.io/?${website}`)
         .then( responseData => {
           let data = responseData.data;
           const parser = new DOMParser();
           const doc = parser.parseFromString(data, "text/html");

           this.setState({ doc });

           const tags = {};
           self.addTags(doc, tags);

           let sortedTags = Object.keys(tags).sort( (a, b) => (tags[b]-tags[a]));

           this.setState({ tags, error: '', newTags: true, sortedTags, data });
         })
         .catch( error => {
           console.log(error);
           this.setState({
             error: 'Unable to fetch the HTML for this page. Please input the full address for this web page.',
             tags: {},
             newTags: true,
             sortedTags: {},
             data: ''
           })
         });
  }

  updateWebsite(e) {
    this.setState({ website: e.target.value, newTags: false });
  }

  addTags(doc, tags) {
    const children = doc.children;
    const len = children.length;
    if (len <= 0) return;

    for (let i = 0; i < len; i++) {
      const currentTag = children[i].tagName;

      tags[currentTag] = tags[currentTag] ? tags[currentTag] + 1 : 1;

      this.addTags(children[i], tags);
    }
  }

  handleMark(e) {
    let tag = e.target.innerText.match(/(.+):/)[1];
    let regex1 = new RegExp(`<${tag}( [^>]*>|>)`, "g");
    let regex2 = new RegExp(`</${tag}>`, "g");
    let mark = new window.Mark(".data-section");
    let targetVal = e.target.classList.value;

    if (e.target.classList.value.slice(0, 6) !== "marked") {
      e.target.classList.value = "marked " + e.target.classList.value;

      mark.markRegExp(regex1);
      mark.markRegExp(regex2);
      let newRegex = [regex1, regex2];
      this.setState({ regex: this.state.regex.concat(newRegex) });

    } else {
      e.target.classList.value = e.target.classList.value.slice(7);
      // mark.unmark({ element: `${tag}` });
      mark.unmark();

      let reg = this.state.regex;

      for (let i = 0; i < reg.length; i++) {
        let idx = String(reg[i]).indexOf(`${tag}`);
        if (idx === -1) {
          mark.markRegExp(reg[i]);
        } else {
          reg.splice(i, 1);
          i--;
        }
      }

      this.setState({regex: reg });
    }
  }

  render() {
    const { tags, error, newTags, sortedTags, data } = this.state;

    return(
      <div className="website-search">
        <div className="top-nav">
          <img src="https://res.cloudinary.com/sharebnb/image/upload/v1505718841/favicon_nysktn.png" className="fetch-logo" alt="dog-logo" />
          <form className="website-form" onSubmit={this.handleSubmit} onChange={this.updateWebsite}>
            <label>Search website:
              <input type="text" name="website" className="search"></input>
            </label>

            <input type="submit" value="Search"></input>
          </form>

        </div>

        <p className="error">{error}</p>

        <TagInfo tags={tags} newTags={newTags} sortedTags={sortedTags} handleMark={this.handleMark} />

        <p className="data-section">{data}</p>
      </div>
    )
  }
}

export default WebsiteSearch;
