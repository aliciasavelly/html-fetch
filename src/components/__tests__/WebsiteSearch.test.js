import React from 'react';
import TagInfo from '../TagInfo';
import WebsiteSearch from '../WebsiteSearch';
import { shallow, mount } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

describe('WebsiteSearch', () => {
  const shallowComponent = shallow(<WebsiteSearch />);
  const component = mount(<WebsiteSearch />);

  it('includes TagInfo component', () => {
    expect(component.find(TagInfo).length).toBe(1);
  });

  it('shows website search section', () => {
    const tree = shallowToJson(shallowComponent);

    expect(tree).toMatchSnapshot();
  });

  it('changes website prop when user types in search bar', () => {
    const websiteInput = component.find('input[type="text"]');
    const site = 'https://aliciasavelly.github.io/portfolio/';
    websiteInput.simulate('change', { target: { value: site }});

    expect(component.state().website).toEqual(site);
  });
});
