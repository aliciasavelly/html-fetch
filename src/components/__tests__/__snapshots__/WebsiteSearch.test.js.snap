// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WebsiteSearch shows website search section 1`] = `
<div
  className="website-search"
>
  <div
    className="top-nav"
  >
    <img
      alt="dog-logo"
      className="fetch-logo"
      src="https://res.cloudinary.com/sharebnb/image/upload/v1505718841/favicon_nysktn.png"
    />
    <form
      className="website-form"
      id="form"
      onChange={[Function]}
      onSubmit={[Function]}
    >
      <label>
        Search website:  
        <input
          className="search"
          name="website"
          type="text"
        />
      </label>
      <input
        id="input-button"
        type="submit"
        value="Search"
      />
    </form>
  </div>
  <p
    className="error"
  />
  <TagInfo
    handleMark={[Function]}
    newTags={false}
    sortedTags={Array []}
    tags={Object {}}
  />
  <p
    className="data-section"
  />
</div>
`;
